Output type of warnings.catch_warnings with record=True

Although [warnings.catch_warnings](https://docs.python.org/3/library/warnings.html#warnings.catch_warnings) with `record=True` returns
a list, mypy is not aware of that:

File `foo.py`:

```python
import warnings

with warnings.catch_warnings(record=True) as w:
    print(len(w))
```

Command `mypy foo.py` gives:

```
$ mypy foo.py
foo.py:4: error: Argument 1 to "len" has incompatible type "Optional[List[_Record]]"; expected "Sized"
Found 1 error in 1 file (checked 1 source file)
```

Using recent mypy from master (`0.750+dev.e97377c454a1d5c019e9c56871d5f229db6b47b2`).

---

I've attached PR fixing this #3464.

--- a/stdlib/2and3/warnings.pyi
+++ b/stdlib/2and3/warnings.pyi
@@ -1,7 +1,13 @@
 # Stubs for warnings
 
-from typing import Any, Dict, List, NamedTuple, Optional, overload, TextIO, Tuple, Type, Union
-from types import ModuleType, TracebackType
+import sys
+from typing import Any, Dict, List, NamedTuple, Optional, overload, TextIO, Tuple, Type, Union, ContextManager
+from types import ModuleType
+
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
 
 @overload
 def warn(message: str, category: Optional[Type[Warning]] = ..., stacklevel: int = ...) -> None: ...
@@ -37,10 +43,12 @@ class _Record(NamedTuple):
     file: Optional[TextIO]
     line: Optional[str]
 
-class catch_warnings:
-    def __init__(self, *, record: bool = ...,
-                 module: Optional[ModuleType] = ...) -> None: ...
-    def __enter__(self) -> Optional[List[_Record]]: ...
-    def __exit__(self, exc_type: Optional[Type[BaseException]],
-                 exc_val: Optional[BaseException],
-                 exc_tb: Optional[TracebackType]) -> None: ...
+
+@overload
+def catch_warnings(*, record: Literal[False] = ..., module: Optional[ModuleType] = ...) -> ContextManager[None]: ...
+
+@overload
+def catch_warnings(*, record: Literal[True], module: Optional[ModuleType] = ...) -> ContextManager[List[_Record]]: ...
+
+@overload
+def catch_warnings(*, record: bool, module: Optional[ModuleType] = ...) -> ContextManager[Optional[List[_Record]]]: ...

